etc/i2c/eeprom.py

```diff
@@ -18,18 +19,24 @@ def read_bytes(bus, device, length = 32):
 # access methods.  In particular, the 24c64 needs a two byte read address to be
 # written to it, whereas the 24c02 interprets a two byte write as writing a byte
 # to the EEPROM.
-#    The code below relies on undocumented behaviour that we are seeing on the
-# M24C64: writing a single byte appears to reset the entire page address (32 bit
-# transfer pages), but not the byte offset within the page.
-#    Therefore the trick here is to read half of the same page twice using
-# single byte addressing: if we get the same result both times then assume that
-# this is ok, otherwise fall back to two byte addressing.
+# Here we attempt to read the IPMI header twice in 8-bit mode.
+# If we get the header successfully twice then 8-bit addressing worked.
+# If not we assume the device is 16-bit addressed 
+# NOTE: this fails for an 8-bit device which doesn't contain valid data
+
 def detect_16bit(device = 0x50):
-    # Read two half pages.  If they are the same then we can assume that we were
-    # able to reset
-    first_read  = read_8bit_address(device, length = 15)
-    second_read = read_8bit_address(device, length = 15)
-    return first_read != second_read
+    first_read  = read_8bit_address(device, length = 8)
+    second_read = read_8bit_address(device, length = 8)
+    try:
+        parse_ipmi.parse_header(first_read)
+        parse_ipmi.parse_header(second_read)
+    except AssertionError:
```

Comment:
Not happy about specifically catching `AssertionError`.  I'd suggest either catching all exceptions (if that's what you want), or hunting down the offending `assert` statements and replacing them with raises of an appropriate new `Exception` subclass.

Reply:
the asserts are inside parse_header....maybe a ParsingException?